src/internal_utils.jl

```diff
@@ -0,0 +1,25 @@
+"""
+This file contains utility functions that are not specific to the
+library, but are used throughout.
+"""
+
+@generated function fieldnames_equal(::Type{T1}, ::Type{T2}) where {T1,T2}
+    # Needs to be a generated function as might not be inlined
```

Comment:
I'm not sure if "inlined" is the right word here, maybe better to simply say "to ensure the check occurs at compile time"?